recorder.go

```diff
@@ -47,51 +52,79 @@ func NewRecorder(ctx context.Context, opts ...Option) (*Recorder, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &Recorder{
+
+	rec := &Recorder{
 		project:     options.projectID,
 		ctx:         ctx,
 		traceClient: c,
 		log:         options.log,
-	}, nil
+	}
+
+	bundler := bundler.NewBundler((*pb.Trace)(nil), func(bundle interface{}) {
+		traces := bundle.([]*pb.Trace)
+		err := rec.upload(traces)
+		if err != nil {
+			log.Printf("failed to upload %d traces to the Cloud Trace server. (err = %s)", len(traces), err)
+		}
+	})
+	bundler.DelayThreshold = 2 * time.Second
+	bundler.BundleCountThreshold = 100
+	// We're not measuring bytes here, we're counting traces and spans as one "byte" each.
+	bundler.BundleByteThreshold = 1000
+	bundler.BundleByteLimit = 1000
+	bundler.BufferedByteLimit = 10000
+	rec.bundler = bundler
+
+	return rec, nil
 }
 
 // RecordSpan writes Span to the GCLoud StackDriver.
 func (r *Recorder) RecordSpan(sp basictracer.RawSpan) {
 	traceID := fmt.Sprintf("%016x%016x", sp.Context.TraceID, sp.Context.TraceID)
 	nanos := sp.Start.UnixNano()
+	labels := convertTags(sp.Tags)
+	transposeLabels(labels)
+	addLogs(labels, sp.Logs)
 
-	req := &pb.PatchTracesRequest{
+	trace := &pb.Trace{
 		ProjectId: r.project,
-		Traces: &pb.Traces{
-			Traces: []*pb.Trace{
-				{
-					ProjectId: r.project,
-					TraceId:   traceID,
-					Spans: []*pb.TraceSpan{
-						{
-							SpanId: sp.Context.SpanID,
-							Kind:   convertSpanKind(sp.Tags),
-							Name:   sp.Operation,
-							StartTime: &timestamp.Timestamp{
-								Seconds: nanos / 1e9,
-								Nanos:   int32(nanos % 1e9),
-							},
-							EndTime: &timestamp.Timestamp{
-								Seconds: (nanos + int64(sp.Duration)) / 1e9,
-								Nanos:   int32((nanos + int64(sp.Duration)) % 1e9),
-							},
-							ParentSpanId: sp.ParentSpanID,
-							Labels:       convertTags(sp.Tags),
-						},
-					},
+		TraceId:   traceID,
+		Spans: []*pb.TraceSpan{
+			{
+				SpanId: sp.Context.SpanID,
+				Kind:   convertSpanKind(sp.Tags),
+				Name:   sp.Operation,
+				StartTime: &timestamp.Timestamp{
+					Seconds: nanos / 1e9,
+					Nanos:   int32(nanos % 1e9),
+				},
+				EndTime: &timestamp.Timestamp{
+					Seconds: (nanos + int64(sp.Duration)) / 1e9,
+					Nanos:   int32((nanos + int64(sp.Duration)) % 1e9),
 				},
+				ParentSpanId: sp.ParentSpanID,
+				Labels:       labels,
 			},
 		},
 	}
+	go func() {
+		err := r.bundler.Add(trace, 2) // size = (1 trace + 1 span)
+		if err == bundler.ErrOversizedItem {
+			log.Printf("trace upload bundle too full. uploading immediately")
+			err = r.upload([]*pb.Trace{trace})
+			if err != nil {
+				log.Println("error uploading trace:", err)
```

Comment:
Can you please use here logger from Recorder struct? I mean `r.log.Errorf(....)`